File: add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode current = dummy;
        int carry = 0;
        
        while(l1 != null || l2 != null){
            int one = l1 != null ? l1.val : 0;
            int two = l2 != null ? l2.val : 0;
            int sum = carry + one + two;
            current.next = new ListNode(sum%10);
            carry = sum/10;
            if(l1 != null) l1 = l1.next;
            if(l2 != null) l2 = l2.next;
            current = current.next;
        }
        
        if(carry > 0) current.next = new ListNode(carry);
        
        return dummy.next;
    }
}

File: add-two-numbers/add-two-numbers.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val dummy = ListNode(-1)
        var current = dummy 
        var tmpOne = l1
        var tmpTwo = l2
        
        var carry = 0
        while(tmpOne != null || tmpTwo != null){
            val one = if (tmpOne != null) tmpOne.`val` else 0
            val two = if (tmpTwo != null) tmpTwo.`val` else 0
            val sum = one + two + carry
            carry = sum / 10
            println(carry)
            current.next = ListNode(sum % 10)
            current = current.next
            if(tmpOne != null)tmpOne = tmpOne.next
            if(tmpTwo != null)tmpTwo = tmpTwo.next
        }
        
        if(carry > 0) current.next = ListNode(carry)
        
        return dummy.next
    }
}